src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -448,19 +462,20 @@ private void OnProjectAdded(ProjectInfo projectInfo, bool silent)
             }
         }
 
-        private void OnProjectAdded_NoLock(ProjectInfo projectInfo, bool silent)
+        private (Solution oldSolution, Solution newSolution) OnProjectAdded_NoLock(ProjectInfo projectInfo, bool silent)
         {
             var projectId = projectInfo.Id;
 
             CheckProjectIsNotInCurrentSolution(projectId);
 
-            var oldSolution = this.CurrentSolution;
-            var newSolution = this.SetCurrentSolution(oldSolution.AddProject(projectInfo));
+            var (oldSolution, newSolution) = this.SetCurrentSolution(this.CurrentSolution.AddProject(projectInfo));
 
             if (!silent)
             {
                 this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.ProjectAdded, oldSolution, newSolution, projectId);
             }
```

Comment:
an alternate to this pattern would be to never raise events here, and have htat happen in the callers who pass 'true'.